compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -4973,6 +4973,14 @@ object Types {
         foldOver(n + 1, tp)
       case tp: TypeRef if tp.info.isTypeAlias =>
         apply(n, tp.superType)
+      case tp: TermRef =>
+        apply(n, tp.underlying)
+      case tp: TypeParamRef =>
+        ctx.typerState.constraint.entry(tp) match {
+          case tb: TypeBounds => foldOver(n, tb)
```

Comment:
This case seems to be redundant. Is this not covered by the third case?
Also, what if the type is F-bounded? Would that not give an infinite recursion?

Reply:
The type being F-bounded _is_ a problem. But we do need to look into those bounds. What are the options here?